File: src/main/java/org/genyris/interp/PairSourceEnvironment.java
package org.genyris.interp;

import org.genyris.core.*;
import org.genyris.exception.GenyrisException;

public class PairSourceEnvironment extends PairEnvironment {
    public PairSourceEnvironment(Environment runtime, Pair theObject)
            throws GenyrisException {
        super(runtime, theObject);

    }

    public Exp lookupDynamicVariableValue(DynamicSymbol dsym)
            throws UnboundException {
        Symbol sym = dsym.getRealSymbol();
        if (sym == _lineNumber) {
            return new Bignum(((PairSource)_theExpression).lineNumber);
        } else if (sym == _filename) {
            return new StrinG(((PairSource)_theExpression).filename);
        } else {
            return super.lookupDynamicVariableValue(dsym);
        }
    }

    public Exp lookupLexicalVariableValue(SimpleSymbol sym)
			throws UnboundException {
        if (sym == _lineNumber) {
            return new Bignum(((PairSource)_theExpression).lineNumber);
        } else if (sym == _filename) {
            return new StrinG(((PairSource)_theExpression).filename);
        } else {
            return super.lookupLexicalVariableValue(sym);
        }
	}


}

File: src/org/genyris/interp/PairSourceEnvironment.kt
package org.genyris.interp

import org.genyris.core.*

class PairSourceEnvironment(runtime: Environment?, theObject: Pair?) : PairEnvironment(runtime, theObject) {
    @kotlin.Throws(UnboundException::class)
    override fun lookupDynamicVariableValue(dsym: DynamicSymbol): Exp? {
        val sym = dsym.getRealSymbol()
        if (sym === _lineNumber) {
            return Bignum((_theExpression as PairSource).lineNumber)
        } else if (sym === _filename) {
            return StrinG((_theExpression as PairSource).filename)
        } else {
            return super.lookupDynamicVariableValue(dsym)
        }
    }

    @kotlin.Throws(UnboundException::class)
    override fun lookupLexicalVariableValue(sym: SimpleSymbol?): Exp? {
        if (sym === _lineNumber) {
            return Bignum((_theExpression as PairSource).lineNumber)
        } else if (sym === _filename) {
            return StrinG((_theExpression as PairSource).filename)
        } else {
            return super.lookupLexicalVariableValue(sym)
        }
    }
}
